scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -250,25 +250,52 @@ def highly_variable_genes(
             'pass `inplace=False` if you want to return a `np.recarray`.')
 
     if batch_key is None:
-        df = _highly_variable_genes_single_batch(adata,
-                                                 min_disp=min_disp, max_disp=max_disp,
-                                                 min_mean=min_mean, max_mean=max_mean,
-                                                 n_top_genes=n_top_genes,
-                                                 n_bins=n_bins,
-                                                 flavor=flavor)
+        df = _highly_variable_genes_single_batch(
+            adata,
+            min_disp=min_disp, max_disp=max_disp,
+            min_mean=min_mean, max_mean=max_mean,
+            n_top_genes=n_top_genes,
+            n_bins=n_bins,
+            flavor=flavor,
+        )
     else:
         sanitize_anndata(adata)
         batches = adata.obs[batch_key].cat.categories
         df = []
         for batch in batches:
             adata_subset = adata[adata.obs[batch_key] == batch]
-            hvg = _highly_variable_genes_single_batch(adata_subset,
-                                                      min_disp=min_disp, max_disp=max_disp,
-                                                      min_mean=min_mean, max_mean=max_mean,
-                                                      n_top_genes=n_top_genes,
-                                                      n_bins=n_bins,
-                                                      flavor=flavor)
-            hvg['gene'] = adata.var_names.values
+
+            # Filter to genes that are in the dataset
+            genes_before = adata_subset.var_names
+            filter_genes(adata_subset, min_cells=1)
+            genes_after = adata_subset.var_names
+            genes_missing = genes_before.difference(genes_after)
+
+            hvg = _highly_variable_genes_single_batch(
+                adata_subset,
+                min_disp=min_disp, max_disp=max_disp,
+                min_mean=min_mean, max_mean=max_mean,
+                n_top_genes=n_top_genes,
+                n_bins=n_bins,
+                flavor=flavor,
+            )
+
+            # Add 0 values for genes that were filtered out
+            missing_hvg = pd.DataFrame(
+                np.zeros((len(genes_missing), len(hvg.columns))),
+                columns=hvg.columns,
+            )
+            missing_hvg['highly_variable'] = missing_hvg['highly_variable'].astype(bool)
+            missing_hvg['gene'] = genes_missing
+            hvg['gene'] = adata_subset.var_names.values
+            hvg = hvg.append(missing_hvg, ignore_index=True)
+
+            # Order as before filtering
+            tmp = hvg.set_index('gene')
+            tmp = tmp.loc[genes_before]
+            tmp['gene'] = tmp.index
+            tmp.index = hvg.index
```

Comment:
Do these lines end up doing anything?

Reply:
They didn't before... I had forgotten to change the line after (`df.append(hvg)` is now `df.append(tmp)`). Thanks for the heads up. As I mentioned, I have only tested that it runs so far.